## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [89]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [90]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [91]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [92]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [93]:
# What three item categories had the most entries?
df.category.value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [94]:
# For the category with the most entries, which subcategory had the most entries?
df.loc[df.category == 'consumables', 'subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [95]:
# Which five clients had the most entries in the data?
df.client_id.value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [96]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df.client_id.value_counts().head(5).index.tolist()
top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [97]:
# How many total units (the qty column) did the client with the most entries order order?

df.loc[df.client_id == 33615, 'qty'].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [98]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df[['unit_price', 'qty', 'subtotal']].head()

unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [99]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def shipping_calculate(total_weight):
    if total_weight > 50:
        return total_weight * 7
    else:
        return total_weight * 10

        
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(shipping_calculate)
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [100]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.095
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  total_weight  shipping_price  \
0         7.50  105            1  115164.00        787.50         5512.50   
1         1.49   21            0     523.95         31.29          312.90   
2         1.68   39            6     527.28         65.52          458.64   
3         1.23   29            3    1056.18         35.67          356.70   
4        46.43   20            1    3902.00        928.60         6500.20   

    total_price  
0  132140.76750  
1     916.35075  
2    1079.58240  
3    1547.10360  
4   11390.40900  

[5 rows x 22 columns]

In [101]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  total_weight  shipping_price   total_price  \
0            1  115164.00        787.50         5512.50  132140.76750   
1            0     523.95         31.29          312.90     916.35075   
2            6     527.28         65.52          458.64    1079.58240   
3            3    1056.18         35.67          356.70    1547.10360   
4            1    3902.00        928.60         6500.20   11390.40900   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [102]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = (df['total_price'] - df['line_cost'])
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price   total_price  line_cost  \
0  115164.00        787.50         5512.50  132140.76750   85597.05   
1     523.95         31.29          312.90     916.35075     629.79   
2     527.28         65.52          458.64    1079.58240     765.18   
3    1056.18         35.67          356.70    1547.10360    1077.35   
4    3902.00        928.60         6500.20   11390.40900    8663.60   

        profit  
0  46543.71750  
1    286.56075  
2    314.40240  
3    469.75360  
4   2726.80900  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [103]:
# Check your work using the totals above
# List of specific order IDs to calculate the total price for
order_ids = [2742071, 2173913, 6128929]

# Filter the DataFrame for only the selected order IDs and make a copy
selected_orders = df[df['order_id'].isin(order_ids)].copy()

# Sum the 'subtotal' and 'shipping_price' per order_id to get the total for each order
order_totals = selected_orders.groupby('order_id').agg({
    'subtotal': 'sum',
    'shipping_price': 'sum'
}).reset_index()

# Calculate the total price for each of these orders using the formula
order_totals['total_price'] = (order_totals['subtotal'] + order_totals['shipping_price']) * 1.0925

# Display the order IDs and their corresponding total prices
print(order_totals[['order_id', 'total_price']])

# Verify with the receipts
expected_prices = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Compare the calculated total prices with the expected values
for order_id in order_ids:
    calculated = order_totals[order_totals['order_id'] == order_id]['total_price'].values[0]
    expected = expected_prices[order_id]
    print(f"Order ID {order_id}:")
    print(f"Calculated Total Price: ${calculated:.2f}")
    print(f"Expected Total Price: ${expected:.2f}")
    print(f"Difference: ${calculated - expected:.2f}\n")



   order_id    total_price
0   2173913  162388.719300
1   2742071  152811.897075
2   6128929  923441.240450
Order ID 2742071:
Calculated Total Price: $152811.90
Expected Total Price: $152811.89
Difference: $0.01

Order ID 2173913:
Calculated Total Price: $162388.72
Expected Total Price: $162388.71
Difference: $0.01

Order ID 6128929:
Calculated Total Price: $923441.24
Expected Total Price: $923441.25
Difference: $-0.01



## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [104]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
# Step 1: Filter the DataFrame for only the top 5 clients
top_clients_data = df[df['client_id'].isin(top_5_clients)]

# Step 2: Calculate the total spent by each client (sum of total price per client)
client_spending = top_clients_data.groupby('client_id')['total_price'].sum().reset_index()

# Step 3: Print the total spending by each of the top 5 clients
print(client_spending[['client_id', 'total_price']])



   client_id   total_price
0      24741  8.245715e+07
1      33615  8.396479e+06
2      38378  1.293609e+07
3      46820  9.766091e+06
4      66037  1.028299e+07


In [105]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Step 1: Filter the DataFrame for only the top 5 clients
top_clients_data = df[df['client_id'].isin(top_5_clients)]

# Step 2: Calculate the required totals for each client
summary = top_clients_data.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
).reset_index()

# Step 3: Print the summary DataFrame
print(summary)


   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      24741                 239862            5126448.37   8.245715e+07   
1      33615                  64313            1828984.89   8.396479e+06   
2      38378                  73667            3429455.40   1.293609e+07   
3      46820                  75768            1601448.84   9.766091e+06   
4      66037                  43018            1395151.85   1.028299e+07   

   total_profit  
0  3.676825e+07  
1  2.221165e+06  
2  3.301364e+06  
3  2.758900e+06  
4  3.278509e+06  


In [106]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns =['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return f"${x / 1_000_000:.2f}M" if pd.notnull(x) else "$0.00M"

# Apply the currency_format_millions function to only the money columns. 
summary[money_columns] = summary[money_columns].applymap(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
summary.rename(columns={
               'total_units_purchased': 'Total Units Purchased',
               'total_shipping_price': 'Total Shipping Price (in millions)',
               'total_revenue': 'Total Revenue (in millions)',
               'total_profit': 'Total Profit (in millions)'
               }, inplace=True)
print(summary)



   client_id  Total Units Purchased Total Shipping Price (in millions)  \
0      24741                 239862                             $5.13M   
1      33615                  64313                             $1.83M   
2      38378                  73667                             $3.43M   
3      46820                  75768                             $1.60M   
4      66037                  43018                             $1.40M   

  Total Revenue (in millions) Total Profit (in millions)  
0                     $82.46M                    $36.77M  
1                      $8.40M                     $2.22M  
2                     $12.94M                     $3.30M  
3                      $9.77M                     $2.76M  
4                     $10.28M                     $3.28M  


C:\Users\jchay\AppData\Local\Temp\ipykernel_51780\3404922066.py:11: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary[money_columns] = summary[money_columns].applymap(currency_format_millions)


In [108]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary = summary.sort_values(by='Total Profit (in millions)', ascending=False)
print(sorted_summary)

   client_id  Total Units Purchased Total Shipping Price (in millions)  \
0      24741                 239862                             $5.13M   
2      38378                  73667                             $3.43M   
4      66037                  43018                             $1.40M   
3      46820                  75768                             $1.60M   
1      33615                  64313                             $1.83M   

  Total Revenue (in millions) Total Profit (in millions)  
0                     $82.46M                    $36.77M  
2                     $12.94M                     $3.30M  
4                     $10.28M                     $3.28M  
3                      $9.77M                     $2.76M  
1                      $8.40M                     $2.22M  


In [ ]:
#write a brief summary of your findings
#During this exercise I went through what appears to be a companys order management system. I helped indentify their top 5 clients total spend in millions, what their most popular products being ordered are and also their profit margins. With this code a company could now understand future forcasting demands and indentify top priority retention clients. 